// react-native-simple-sticky-header/lib/components/Button.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Button = void 0;
const react_1 = __importDefault(require("react"));
const react_native_1 = require("react-native");
function Button({ children, active = false }) {
    return (react_1.default.createElement(react_native_1.TouchableOpacity, null,
        react_1.default.createElement(react_native_1.View, { style: {
                borderRadius: 25,
                backgroundColor: active ? '#333333' : '#ffffff',
                paddingVertical: 12,
                paddingHorizontal: 16,
                marginHorizontal: 4,
                width: 116,
            } },
            react_1.default.createElement(react_native_1.Text, { style: {
                    color: active ? '#ffffff' : '#333333',
                    textAlignVertical: 'center',
                    textAlign: 'center',
                } }, children))));
}
exports.Button = Button;
