// react-navigation-collapsible/lib/src/utils.js
import { Platform, StatusBar } from 'react-native';
import { isIphoneX } from 'react-native-iphone-x-helper';
var isExpo = false;
try {
    // eslint-disable-next-line no-undef, @typescript-eslint/no-var-requires
    var Constants = require('expo-constants').default;
    // True if the app is running in an `expo build` app or if it's running in Expo Go.
    isExpo =
        Constants.executionEnvironment === 'standalone' ||
            Constants.executionEnvironment === 'storeClient';
    // eslint-disable-next-line no-empty
}
catch (_a) { }
var SAFEBOUNCE_HEIGHT_IOS = 300;
var SAFEBOUNCE_HEIGHT_ANDROID = 100;
var SAFEBOUNCE_HEIGHT_WEB = 0;
var safeBounceHeight = Platform.select({
    ios: SAFEBOUNCE_HEIGHT_IOS,
    android: SAFEBOUNCE_HEIGHT_ANDROID,
    web: SAFEBOUNCE_HEIGHT_WEB,
});
var setSafeBounceHeight = function (height) {
    safeBounceHeight = height;
};
var getSafeBounceHeight = function () { return safeBounceHeight; };
var getDefaultHeaderHeight = function (isLandscape) {
    if (Platform.OS === 'ios') {
        if (isLandscape && !Platform.isPad) {
            return 32;
        }
        else {
            return 44;
        }
    }
    else if (Platform.OS === 'android') {
        return 56;
    }
    return 64;
};
var disabledExpoTranslucentStatusBar = false;
var disableExpoTranslucentStatusBar = function () {
    disabledExpoTranslucentStatusBar = true;
};
var getStatusBarHeight = function (isLandscape) {
    if (Platform.OS === 'ios') {
        if (isLandscape)
            return 0;
        return isIphoneX() ? 44 : 20;
    }
    else if (Platform.OS === 'android') {
        // eslint-disable-next-line no-undef
        return (global.Expo || isExpo) && !disabledExpoTranslucentStatusBar
            ? StatusBar.currentHeight
            : 0;
    }
    else
        return 0;
};
var getNavigationHeight = function (isLandscape, headerHeight) {
    return headerHeight + getStatusBarHeight(isLandscape);
};
var getElevationStyle = function (elevation) {
    if (Platform.OS === 'ios') {
        if (elevation === 0)
            return null;
        else
            return {
                shadowOpacity: 0.0015 * elevation + 0.18,
                shadowRadius: 0.54 * elevation,
                shadowOffset: {
                    height: 0.6 * elevation,
                    width: 0.6 * elevation,
                },
            };
    }
    else {
        return {
            elevation: elevation,
        };
    }
};
export { setSafeBounceHeight, getSafeBounceHeight, getDefaultHeaderHeight, getNavigationHeight, getStatusBarHeight, disableExpoTranslucentStatusBar, getElevationStyle, };
