// react-native-simple-sticky-header/lib/__tests__/StickyHeader.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable @typescript-eslint/unbound-method */
const react_1 = __importDefault(require("react"));
require("react-native");
const react_native_1 = require("@testing-library/react-native");
const StickyHeader_1 = require("../components/StickyHeader");
const react_native_2 = require("react-native");
describe('StickyHeader', () => {
    it('Snapshot Test', () => {
        const { toJSON } = (0, react_native_1.render)(react_1.default.createElement(StickyHeader_1.StickyHeader, { onLayout: jest.fn() },
            react_1.default.createElement(react_native_2.View, { style: {
                    display: 'flex',
                    flexDirection: 'row',
                    justifyContent: 'space-around',
                    width: '100%',
                    alignItems: 'center',
                    height: 64,
                } },
                react_1.default.createElement(react_native_2.View, { style: {
                        marginTop: 2,
                        width: 16,
                        height: 16,
                    } }),
                react_1.default.createElement(react_native_2.View, null,
                    react_1.default.createElement(react_native_2.TouchableOpacity, null,
                        react_1.default.createElement(react_native_2.Text, { style: {
                                fontSize: 16,
                            } }, "\u4ECA\u3059\u3050\u30FB\u6771\u4EAC\u90FD\u5343\u4EE3\u7530\u533A1\u4E01\u76EE1\u756A\u5730"))),
                react_1.default.createElement(react_native_2.View, { style: {
                        marginTop: 2,
                    } },
                    react_1.default.createElement(react_native_2.TouchableOpacity, null,
                        react_1.default.createElement(react_native_2.Text, { style: {
                                fontSize: 16,
                            } }, "\uD83D\uDC7E"))))));
        expect(toJSON()).toMatchSnapshot();
    });
});
